pullman/storageproviders/gcs/downloader.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2021 IBM Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package gcsprovider
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	"cloud.google.com/go/storage"
+	"github.com/go-logr/logr"
+	"github.com/kserve/modelmesh-runtime-adapter/pullman"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+type gcsClientFactory struct{}
+
+// gcsClientFactory implements gcsDownloaderFactory
+var _ gcsDownloaderFactory = (*gcsClientFactory)(nil)
+
+func (f gcsClientFactory) newDownloader(log logr.Logger, credentials map[string]string) (gcsDownloader, error) {
+	ctx := context.Background()
+	var cl *storage.Client
+	var err error
+
+	if len(credentials) > 0 {
+		credJson, marshalErr := json.Marshal(credentials)
+		if marshalErr != nil {
+			return nil, fmt.Errorf("unable to marshal json credentials: %w", marshalErr)
+		}
+		cl, err = storage.NewClient(ctx, option.WithCredentialsJSON(credJson))
+	} else {
+		cl, err = storage.NewClient(ctx, option.WithoutAuthentication())
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &gcsImplDownloader{
+		client: cl,
+		log:    log,
+	}, nil
+}
+
```

Comment:
Add "interface assertion" for this one too?
```go
var _ gcsDownloader = (*gcsImplDownloader)(nil)
```

Reply:
It's not a requirement... I just find it more readable if the implemented interfaces are made explicit.